### PR TITLE
Use `fcntl(fd, F_GETFD)` to detect if standard streams are open

### LINKED ISSUE #96621: Standard file descriptor checks cause abort on startup with strict rlimit
### Cause

[PR 75295](https://github.com/rust-lang/rust/pull/75295) added polls to the standard file descriptors on startup; if they are closed, it attempts to reopen them to `/dev/null`.  If [`RLIMIT_NOFILE`](https://man7.org/linux/man-pages/man2/getrlimit.2.html) is too low, the polling (or opening) will cause an abort.

[Reported by @edre on URLO; diagnosed by @cuviper.](https://users.rust-lang.org/t/trivial-rust-binaries-no-longer-run-with-rlimit-nofile-0-after-rust-1-48/74775)

The `rlimit` could be checked ahead of time.

### Code

I tried this Docker configuration

```
FROM rust:1.48
RUN rustup target add x86_64-unknown-linux-musl
RUN echo 'fn main(){println!("hi");}' > hi.rs; rustc --target x86_64-unknown-linux-musl -o hi hi.rs
RUN python3 -c 'import resource,subprocess;print(subprocess.run(["/hi"],preexec_fn=lambda: resource.setrlimit(resource.RLIMIT_NOFILE, (0,0))))'
```

I expected to see this happen: Successful execution

Instead, this happened: Process aborted

### Version it worked on

It most recently worked on: Rust 1.47

### Versions with regression

Rust 1.48 and forward.

### Related issues
* https://github.com/rust-lang/rust/issues/47271
* https://github.com/rust-lang/rust/issues/88825

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged

### PR DESCRIPTION
**Pull request by tmiasko**
In the previous implementation, if the standard streams were open,
but the RLIMIT_NOFILE value was below three, the poll would fail
with EINVAL:

> ERRORS: EINVAL The nfds value exceeds the RLIMIT_NOFILE value.

Switch to the existing fcntl based implementation to avoid the issue.

Fixes #96621.

### CODE REVIEW

**Comment by joshtriplett** [comment]
LGTM.

@bors r+

**Comment by the8472** [comment]
This costs 2 more syscalls at program startup, which adds up to a few microseconds I guess? Not the end of the world but still slower than the previous implementation.

**Comment by joshtriplett** [comment]
@the8472 I would be happy to merge a version that tries `poll` and falls back to `fcntl` iff `poll` returns `EINVAL`.

In this context, I felt like it made sense to merge the new test and the code to make that test pass, and then a subsequent PR can add the fast-path and ensure the test still passes.

**Comment by tmiasko** [comment]
* Disabled checking of a compiler stderr in the test.

**Comment by the8472** [comment]
This was discussed this in today's libs meeting. To keep the amount of unavoidable syscalls at program startup small we'd prefer to keep in the poll as fast path and only fall back to fcntl when it returns the error indicating that the rlimit was hit. 

Since this PR mostly removes code it wouldn't make much sense to add it back with different control flow in a separate PR.

**Comment by joshtriplett** [comment]
@tmiasko Do you have the bandwidth to update this PR to try poll first? (Sorry for the churn here.)

**Comment by the8472** [comment]
I have done the changes and pushed them to tmiasko's branch.

File: library/std/src/sys/unix/mod.rs [hunk 1]
```diff
@@ -67,35 +67,64 @@ pub unsafe fn init(argc: isize, argv: *const *const u8) {
...
+                libc::pollfd { fd: 0, events: 0, revents: 0 },
+                libc::pollfd { fd: 1, events: 0, revents: 0 },
+                libc::pollfd { fd: 2, events: 0, revents: 0 },
+            ];
+
+            while libc::poll(pfds.as_mut_ptr(), 3, 0) == -1 {
+                if errno() == libc::EINTR {
+                    continue;
+                }
+                if errno() == libc::EINVAL {
```
Comment (tmiasko) [inline library/std/src/sys/unix/mod.rs:94]: The `ENOMEM` is another situation where use of the fallback might be appropriate.

Comment (the8472) [reply library/std/src/sys/unix/mod.rs:94]: When/where does that happen? It's not listed in the posix or linux manpages.

Comment (tmiasko) [reply library/std/src/sys/unix/mod.rs:94]: > ENOMEM Unable to allocate memory for kernel data structures.
> ...
> This page is part of release 5.13 of the Linux  man-pages  project.

Comment (tmiasko) [reply library/std/src/sys/unix/mod.rs:94]: But, yes posix suggest a different error code in such a situation:

> [EAGAIN] The allocation of internal data structures failed but a subsequent request may succeed.

File: library/std/src/sys/unix/mod.rs [hunk 2]
```diff
@@ -67,35 +67,65 @@ pub unsafe fn init(argc: isize, argv: *const *const u8) {
...
+            while libc::poll(pfds.as_mut_ptr(), 3, 0) == -1 {
+                if errno() == libc::EINTR {
+                    continue;
+                }
+                if matches!(errno(), libc::EINVAL | libc::EAGAIN | libc::ENOMEM) {
+                    // RLIMIT_NOFILE or temporary allocation failures
+                    // may be preventing use of poll(), fall back to fcntl
+                    break 'poll;
+                }
+                libc::abort();
```
Comment (klensy) [inline library/std/src/sys/unix/mod.rs:99]: can this be match like
```rust
match errno() {
libc::EINTR => continue,
...
}
```
so not to call `errno` additional times and i guess this should look nicer?

Comment (the8472) [reply library/std/src/sys/unix/mod.rs:99]: Good point

**Comment by the8472** [comment]
Converted the ifs to a match, I assume that's trivial enough that it won't need another review.

@bors r=joshtriplett

**Comment by spastorino** [comment]
The test added in this PR now forces to install `glibc.static` in nixos. See https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/process.2Fnofile-limit.2Ers.20test.20failing.20on.20nixos

File: library/std/src/sys/unix/mod.rs [hunk 3]
```diff
@@ -67,45 +66,66 @@ pub unsafe fn init(argc: isize, argv: *const *const u8) {
...
-                while libc::poll(pfds.as_mut_ptr(), 3, 0) == -1 {
-                    if errno() == libc::EINTR {
-                        continue;
+        // fast path with a single syscall for systems with poll()
+        #[cfg(not(any(
+            miri,
+            target_os = "emscripten",
+            target_os = "fuchsia",
+            target_os = "vxworks",
+            // The poll on Darwin doesn't set POLLNVAL for closed fds.
```
Comment (RalfJung) [inline library/std/src/sys/unix/mod.rs:76]: FWIW I cannot reproduce this. Miri runs its poll tests also natively on macOS and we are getting POLLNVAL for bogus FDs as expected.

However we are also getting POLLNVAL on stdin even when that is not closed...

**Pull request merged**